File: Sean的leetcode/FB/Binary Tree Level Order Traversal.py
###
# BFS
# Time Complexity: O(n)
# Space Complexity: O(n)
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        q = collections.deque([root])
        re = []
        while q:
            n = len(q)
            level = []
            for i in xrange(n):
                e = q.popleft()
                level.append(e.val)
                if e.left:
                    q.append(e.left)
                if e.right:
                    q.append(e.right)
            re.append(level)
        return re

# Solution 2
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        q = [root]
        re = []
        while q:
            next_q = []
            level = []
            for n in q:
                level.append(n.val)
                if n.left:
                    next_q.append(n.left)
                if n.right:
                    next_q.append(n.right)
            re.append(level)
            q = next_q
        return re
        

File: MyLeetCode/Leetcode/Binary Tree Level Order Traversal.cpp
#if 1 //iterative solution
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int> > res;
        vector<int> one_row;
        queue<TreeNode *> q;

        if (!root) return res;

        q.push(root);

        while (!q.empty())
        {
            //note: use num to collect the node at the same level.
            int num = q.size();
            for (int i = 0; i < num; i++)
            {
                TreeNode *f = q.front();
                q.pop();
                one_row.push_back(f->val);
                if (f->left)
                    q.push(f->left);
                if (f->right)
                    q.push(f->right);
            }
            res.push_back(one_row);
            one_row.clear();
        }
        return res;
    }
};
#else //recursive solution
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void genLevel(TreeNode *root, vector<vector<int>> &res, int level)
    {
        if (!root) return;

        if (level == res.size())
        {
            vector<int> one_level;
            res.push_back(one_level);
        }

        res[level].push_back(root->val);
        genLevel(root->left, res, level+1);
        genLevel(root->right, res, level+1);
    }

    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int> > res;
        genLevel(root, res, 0);
        return res;
    }
};
#endif